subsys/storage/storage_area/storage_area_flash.c

```diff
@@ -0,0 +1,256 @@
+{
+	const struct storage_area_flash *flash =
+		CONTAINER_OF(area, struct storage_area_flash, area);
+	int rc = sa_flash_valid(flash);
+
+	if (rc != 0) {
+		goto end;
+	}
+
+	off_t rdoff = flash->doffset + (off_t)offset;
+
+	for (size_t i = 0U; i < iovcnt; i++) {
+		rc = flash_read(flash->dev, rdoff, iovec[i].data, iovec[i].len);
+		if (rc != 0) {
+			break;
+		}
+
+		rdoff += iovec[i].len;
+	}
+
+	if (rc != 0) {
+		LOG_DBG("read failed at 0x%lx", rdoff - flash->doffset);
+	}
+end:
+	return rc;
+}
+
+static int sa_flash_write(const struct storage_area_flash *flash,
+			  sa_off_t offset, const uint8_t *data, size_t len)
+{
+	const struct storage_area *area = &flash->area;
+	off_t wroff = flash->doffset + (off_t)offset;
+
+	if ((!STORAGE_AREA_AUTOERASE(area)) || (STORAGE_AREA_FOVRWRITE(area))) {
+		return flash_write(flash->dev, wroff, data, len);
+	}
+
+	int rc = 0;
+
+	while (len != 0U) {
+		const size_t esz = area->erase_size;
+		const size_t wrlen = MIN(esz - (wroff & (esz - 1)), len);
+
+		if ((wroff & (esz - 1)) == 0U) {
+			rc = flash_erase(flash->dev, wroff, esz);
+			if (rc != 0) {
+				break;
+			}
+		}
+
+		rc = flash_write(flash->dev, wroff, data, wrlen);
+		if (rc != 0) {
+			break;
+		}
+
+		wroff += wrlen;
+		data += wrlen;
+		len -= wrlen;
+	}
+
+	if (rc != 0) {
+		LOG_DBG("prog failed at 0x%lx", wroff - flash->doffset);
+	}
+
+	return rc;
+}
+
+static int sa_flash_writev(const struct storage_area *area, sa_off_t offset,
+			   const struct storage_area_iovec *iovec, size_t iovcnt)
+{
+	const struct storage_area_flash *flash =
+		CONTAINER_OF(area, struct storage_area_flash, area);
+	const size_t align = area->write_size;
+	uint8_t buf[align];
+	size_t bpos = 0U;
+	int rc = sa_flash_valid(flash);
+
+	if (rc != 0) {
+		goto end;
+	}
+
+	for (size_t i = 0U; i < iovcnt; i++) {
+		uint8_t *data8 = (uint8_t *)iovec[i].data;
+		size_t blen = iovec[i].len;
+
+		if (bpos != 0U) {
+			size_t cplen = MIN(blen, align - bpos);
+
+			memcpy(buf + bpos, data8, cplen);
+			bpos += cplen;
+			blen -= cplen;
+			data8 += cplen;
+
+			if (bpos == align) {
+				rc = sa_flash_write(flash, offset, buf, align);
+				if (rc != 0) {
+					break;
+				}
+
+				offset += align;
+				bpos = 0U;
+			}
+		}
+
+		if (blen >= align) {
+			size_t wrlen = blen & ~(align - 1);
+
+			rc = sa_flash_write(flash, offset, data8, wrlen);
+			if (rc != 0) {
+				break;
+			}
+
+			blen -= wrlen;
+			data8 += wrlen;
+			offset += wrlen;
+		}
+
+		if (blen > 0U) {
+			memcpy(buf, data8, blen);
+			bpos = blen;
+		}
+	}
+end:
+	return rc;
+}
+
+static int sa_flash_erase(const struct storage_area *area, size_t sblk,
+			  size_t bcnt)
+{
+	const struct storage_area_flash *flash =
+		CONTAINER_OF(area, struct storage_area_flash, area);
+	int rc = sa_flash_valid(flash);
+
+	if (rc != 0) {
+		goto end;
+	}
+
+	const off_t eoff = flash->doffset + sblk * area->erase_size;
+	const size_t esize = bcnt * area->erase_size;
+
+	rc = flash_erase(flash->dev, eoff, esize);
```

Comment:
Not all flash devices support erase. Note also that it will not be required for Flash Devices to have erase as flash api starts to support devices if various capabilities - including lack of erase. This means that you have to have code here checking if erase is supported by device and call flash_flatten for devices that do not have erase.
Generally if you erase here is not mandated by device requirement to erase prior to write, then you can just call flash_flatten without checks, as the function itself will do the data swipe.